src/uwtools/rocoto.py

```diff
@@ -25,20 +35,111 @@ def _add_jobname(tree: dict) -> None:
     """
     values = YAMLConfig(input_yaml)
-    tasks = values["tasks"]
+    tasks = values["workflow"]["tasks"]
     if isinstance(tasks, dict):
         _add_jobname(tasks)
 
+
+def _rocoto_schema_yaml() -> DefinitePath:
+    """
+    The path to the file containing the schema to validate the XML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "rocoto.jsonschema"
+
+
+def _rocoto_schema_xml() -> DefinitePath:
+    """
+    The path to the file containing the Rocoto workflow document template to render.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "schema_with_metatasks.rng"
+
+
+def _write_rocoto_xml(
+    config_file: OptionalPath,
+    rendered_output: DefinitePath,
+) -> None:
+    """
+    Render the given YAML file to XML using the given template.
+
+    :param config_file: Path to YAML input file.
+    :param rendered_output: Path to write rendered XML file.
+    """
+
+    rocoto_template = _rocoto_schema_xml()
+    _add_tasks(config_file)
+
     # Render the template.
-    template = J2Template(values=values.data, template_path=input_template)
-    template.dump(output_path=rendered_output)
+    template = J2Template(values=YAMLConfig(config_file).data, template_path=rocoto_template)
+    template.dump(output_path=str(rendered_output))
+
+
+# Public functions
+def realize_rocoto_xml(
+    config_file: OptionalPath,
+    rendered_output: DefinitePath,
+) -> bool:
+    """
+    Realize the given YAML file to XML, using the Rocoto RelaxNG schema. Validate both the YAML and
+    the XML.
+
+    :param input_yaml: Path to YAML input file.
+    :param rendered_output: Path to write rendered XML file.
+    """
+
+    rocoto_schema = _rocoto_schema_yaml()
+
+    # Validate the YAML.
+    if uwtools.config.validator.validate_yaml(config_file=config_file, schema_file=rocoto_schema):
+        _add_tasks(config_file)
+        # Render the template to a temporary file.
+        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+            _write_rocoto_xml(
+                config_file=config_file,
+                rendered_output=temp_file.name,
+            )
+            # Validate the XML.
+            if validate_rocoto_xml(input_xml=temp_file.name):
+                # If no issues were detected, save temp file and report success.
+                shutil.move(temp_file.name, rendered_output)
+                return True
+        logging.error("Rocoto validation errors identified in %s", temp_file.name)
+        return False
+    logging.error("YAML validation errors identified in %s", config_file)
+    return False
+
+
+def validate_rocoto_xml(input_xml: OptionalPath) -> bool:
+    """
+    Given a rendered XML file, validate it against the Rocoto schema.
+
+    :param input_xml: Path to rendered XML file.
+    """
+    rocoto_template = _rocoto_schema_xml()
```

Comment:
I'll move them inline - honestly, its separated here because I was having an issue where it was not returning the value but trying to call the function and erroring, but that may have been a separate implementation error of mine